gammapy/detect/cwt.py

```diff
@@ -57,200 +58,334 @@ def difference_of_gauss_kernel(radius, scale_step, n_sigmas=8):
+            data.error[idx_scale] = np.sqrt(fftconvolve(total_background, kern ** 2, mode='same'))
 
-        self.approx = fftconvolve(self.counts - self.model - self.background,
+        log.info('Computing approx and approx_bkg')
+        data.approx = fftconvolve(data.counts - data.model - data.background,
                                   self.kern_approx, mode='same')
-        self.approx_bkg = fftconvolve(self.background, self.kern_approx, mode='same')
+        data.approx_bkg = fftconvolve(data.background, self.kern_approx, mode='same')
 
-    def compute_support(self, nsigma=2.0, nsigmap=4.0, remove_isolated=True):
-        """Compute the multiresolution support with hard sigma clipping.
+    def _compute_support(self, data):
+        """
+        Compute the multiresolution support with hard sigma clipping.
 
         Imposing a minimum significance on a connected region of significant pixels
-        (i.e. source detection)
+        (i.e. source detection).
+
+        TODO: document?
+        What's happening here:
+        - calc significance for all scales
+        - for each scale compute support:
+            - create mask, whether significance value more than significance_threshold
+            - find all separated structures on mask, label them (with help
+              of 'scipy.ndimage.label')
+            - for each structure do:
+                - find pixels coordinates of the structure
+                - if just one pixel, we can remove it from mask if we want
+                - if the max value of significance in that structure less
+                  than significance_island_threshold, remove this structure from mask
+            - update support by mask
 
         Parameters
         ----------
-        TODO: document
-
-        Returns
-        -------
-        TODO: for now it's stored as `self.support`. Maybe return
+        data : `~gammapy.detect.CWTData`
+            Images after transform.
         """
         from scipy.ndimage import label
         # TODO: check that transform has been performed
 
-        sig = self.transform / self.error
-
-        for key in self.scales.keys():
-            tmp = sig[key] > nsigma
-            # produce a list of connex structures in the support
-            l, n = label(tmp)
-            for id in range(1, n):
-                index = np.where(l == id)
-                if remove_isolated:
-                    if index[0].size == 1:
-                        tmp[index] *= 0.0  # Remove isolated pixels from support
-                signif = sig[key][index]
-                if signif.max() < nsigmap:  # Remove significant pixels island from support
-                    tmp[index] *= 0.0  # if max does not reach maximal significance
-
-            self.support[key] += tmp
-            self.support[key] = self.support[key] > 0.
-
-        return self.support
-
-    def inverse_transform(self):
+        if data.support is None:
```

Comment:
Maybe also it will be better to create kernel and scales class for avoiding this situation: just to put object of that class as a parameter of CWTData to initialize the data